src/kernel-install/60-ukify.install.in

```diff
@@ -0,0 +1,203 @@
+    p.add_argument('entry_dir',
+                   metavar='ENTRY_DIR',
+                   type=Path,
+                   nargs='?',
+                   help='Type#1 entry directory (ignored)')
+    p.add_argument('kernel_image',
+                   metavar='KERNEL_IMAGE',
+                   nargs='?',
+                   help='Kernel binary')
+    p.add_argument('initrd',
+                   metavar='INITRD…',
+                   nargs='*',
+                   help='Initrd files')
+    p.add_argument('--version',
+                   action='version',
+                   version=f'systemd {__version__}')
+
+    opts = p.parse_args(args)
+
+    if opts.command == 'add':
+        opts.staging_area = Path(os.environ['KERNEL_INSTALL_STAGING_AREA'])
+        path_is_readable(opts.staging_area, dir=True)
+
+    return opts
+
+def we_are_wanted() -> bool:
+    KERNEL_INSTALL_LAYOUT = os.getenv('KERNEL_INSTALL_LAYOUT')
+
+    if KERNEL_INSTALL_LAYOUT != 'uki':
+        log(f'{KERNEL_INSTALL_LAYOUT=}, quitting.')
+        return False
+
+    KERNEL_INSTALL_UKI_GENERATOR = os.getenv('KERNEL_INSTALL_UKI_GENERATOR')
+
+    if KERNEL_INSTALL_UKI_GENERATOR != 'ukify':
+        log(f'{KERNEL_INSTALL_UKI_GENERATOR=}, quitting.')
+        return False
+
+    log('KERNEL_INSTALL_LAYOUT and KERNEL_INSTALL_UKI_GENERATOR are good')
+    return True
+
+def kernel_cmdline() -> str:
+    entry_token = os.environ['KERNEL_INSTALL_ENTRY_TOKEN']
+    machine_id = os.environ['KERNEL_INSTALL_MACHINE_ID']
+
+    if root := os.getenv('KERNEL_INSTALL_CONF_ROOT'):
+        try:
+            options = Path(root).joinpath('cmdline').read_text().split()
+        except FileNotFoundError:
+            return None
+    else:
+        for cmdline, fallback in [
+                ('/etc/kernel/cmdline',     False),
+                ('/usr/lib/kernel/cmdline', False),
+                ('/proc/cmdline',           True),
+        ]:
+            try:
+                options = Path(cmdline).read_text().split()
+            except FileNotFoundError:
+                if fallback:
+                    raise
+
+        if fallback:
+            options = [opt for opt in options
+                       if not opt.startswith(('BOOT_IMAGE=', 'initrd='))]
+
+    # If the boot entries are named after the machine ID, then suffix the kernel
+    # command line with the machine ID we use, so that the machine ID remains
+    # stable, even during factory reset, in the initrd (where the system's machine
+    # ID is not directly accessible yet), and if the root file system is volatile.
+    if (entry_token == machine_id and
+        not any(opt.startswith('systemd.machine_id=') for opt in options)):
+        options += [f'systemd.machine_id={machine_id}']
+
+    cmdline = ' '.join(options)
+    if cmdline.startswith('@'):
+        raise ValueError("Kernel options must not start with '@'")
+    return cmdline
+
+def make_uki(opts, cmdline):
+    output = opts.staging_area / 'uki.efi'
```

Comment:
Can we be a bit smarter about the name? I'm not sure whether it should be in here or in kernel-install itself, but we have a bunch of logic in mkosi to determine the filename of the unified kernel image that I would love to be able to remove in the future (see install_unified_kernel() in mkosi)

Reply:
This name is required by `90-uki-copy.install`. If we want to change this, we should change it in both places. (But I'm not sure we do. The file gets renamed later.)

Reply:
I added a comment.